### ISSUE TITLE
[Win32] Drawing produces unsymmetric results without swt.autoScale=false for 4k monitor with zooming

### ISSUE THREAD

**Issue by tmssngr** [eclipse-platform/eclipse.platform.swt#1596]
**Describe the bug**
Code that draws a circle and and a line symmetric with 100% zoom causes asymmetric results with 175% zoom factor.

**To Reproduce**
Run following snippet
```java
import org.eclipse.swt.*;
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.widgets.*;

public class HiDpiDrawing {
	public static void main(String[] args) {
//		System.getProperties().putIfAbsent("swt.autoScale", "false");
		final Display display = new Display();

		final Shell shell = new Shell(display);
		shell.addListener(SWT.Paint, event -> {
			final GC gc = event.gc;
			gc.setAntialias(SWT.ON);
			gc.setLineWidth(3);
			final Path path = new Path(display);
			int radius = 3;
			path.moveTo(15, 0);
			path.lineTo(15, 15 - radius);
			path.moveTo(15, 15 + radius);
			path.lineTo(15, 100);
			gc.drawPath(path);
			path.dispose();
			gc.setBackground(gc.getForeground());
			drawOval(gc, radius + 2);
			gc.setBackground(new Color(255, 255, 255));
			drawOval(gc, radius);
		});

		shell.setSize(400, 300);
		shell.open();

		while (!shell.isDisposed()) {
			if (!display.readAndDispatch()) {
				display.sleep();
			}
		}

		shell.dispose();
	}

	private static void drawOval(GC gc, int radius) {
		gc.fillOval(15 - radius, 15 - radius, 2 * radius + 1, 2 * radius + 1);
	}
}
```
on Windows with 100% zoom level and with 175% zoom level.

**Expected behavior**
The dot is always drawn symmetric to the line, no matter what zoom level (otherwise it would look ugly).

**Screenshots**
With 100% zoom level:
<img width="23" alt="100" src="https://github.com/user-attachments/assets/fb8f1aff-8e49-4882-9a9f-451ab1b64a7b">
close-up:
<img width="113" alt="100%-zoomed" src="https://github.com/user-attachments/assets/2a77913c-0e28-47bc-bcfe-5a1d6c594f45">
With 175% zoom level:
<img width="33" alt="175" src="https://github.com/user-attachments/assets/13f633cb-998f-4450-b871-9dd9d971e4a8">
close-up:
<img width="271" alt="175%-zoomed" src="https://github.com/user-attachments/assets/656535fe-9fc6-4d62-b1bd-5751081bb6e1">

When removing the `//` from the comment line, the result looks fine for 175% zoom, too.

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [ ] All OS
- - [x] Windows
- - [ ] Linux
- - [ ] macOS 

2. Additional OS info (e.g. OS version, Linux Desktop, etc)
Tried on Windows 11 with 4k monitor.

**Comment by tmssngr** [comment]
With `swt.autoScale=quarter` or `exact` it looks even worse:
<img width="36" alt="175%-swt autoScale=quarter" src="https://github.com/user-attachments/assets/0a53e2ab-7bcd-4057-9584-1feaec2e2853">
close-up:
<img width="226" alt="175%-swt autoScale=quarter-zoomed" src="https://github.com/user-attachments/assets/3e79a8f3-b43e-4fb2-abda-7f2087370c53">
Hence it looks like there is no way to draw symmetric elements without setting `swt.autoScale=false`.

**Comment by HeikoKlare** [comment]
The oval at 100% is also asymmetric (can be seen in different colors from anti-aliasing), it only becomes more visible at 175% because the calculation is more off. With `swt.autoScale=false` you just hide that as rendering just still happens at 100%, where the asymmetry is only harder to see.

**Comment by tmssngr** [comment]
Asking users, I reckon 99% would answer the 100% version looks symmetric.

How do we need to draw these elements that they look fine between 100% and 200% zoom level for 99% of the users?

**Comment by HeikoKlare** [comment]
> Asking users, I reckon 99% would answer the 100% version looks symmetric.

Might be, but what does that change? Using the 175% version and placing users at 175% the distance of the monitor than when using the 100% version, they will probably also consider it symmetric...
It's asymmetric, so there seems to be a bug in rendering, which is more visible when using higher scale values. There is nothing better in using `swt.autoScale=false` than using any other scale value with respect to symmetry. Things are simply just too small to see it when using no scaling for HiDPI monitors.

> How do we need to draw these elements that they look fine between 100% and 200% zoom level for 99% of the users?

My first impression is that this is a bug in SWT. So there may be no other way than looking into SWT code, finding the cause of the bug, and fixing it.

**Comment by tmssngr** [comment]
With `swt.autoScale=false` I have the possibility to optimize it for larger zoom levels, because I have control over all pixels, not just control over double-pixels:

IMHO one problem is that pixel coordinates simply are scaled. But there are draw operations like fillXxx that expect the right/lower coordinate outside the drawn area. The other problem is that for most draw operations only int coordinates are allowed, so it scales only to multiples of 2.

I wonder how other drawing APIs handle such fractional scaling.

**Comment by laeubi** [comment]
If I look at your snippet I'm not sure you are taking the line width into account correctly, you set a line width of three but I can't find you possibly account for this.

Beside that, yes only having int coordinates (except for example Path) is a big limitation compared to e.g. AWTs Graphics2D, e.g AWT uses the concept of Shapes what is really convenient and powerful.

**Comment by tmssngr** [comment]
@laeubi In SmartGit we are using the zoom factor (1 or 2) to draw the lines with the correct width. The result looks correct with all zoom levels.

**Comment by tmssngr** [comment]
Assuming following code:
```java
gc.setLineWidth(1);
gc.drawLine(0, 1, 10, 1);
```
what exactly we can expect to happen for 100% zoom level and especially for 200% zoom level? How wide the line should be with 200% zoom (I'd expect 2 pixels), where exactly it should be drawn, should the line be fully visible if drawn at the coordinate 0?

Should we expect to get the same rendering result on all 3 platforms?

**Comment by tmssngr** [comment]
Please give the pull request a try.

**Comment by tmssngr** [comment]
Yes, this the PR #1607. I know that it can't be merged right now because of feature-freeze, but I want you to give it a try.
